### PR TITLE
proposition for refactoring vertex mapping

### PR DESCRIPTION
**Pull request by timpx**
Gowever we will have to check in the future how subcortical structures are dealt with in the the simulator, because before they were included through the vertex mapping; 
I coudn't check extensively because I got an error when configuring monitors

### CODE REVIEW

File: tvb/simulator/simulator.py [hunk 1]
```diff
@@ -474,7 +471,7 @@ def __call__(self, simulation_length=None, random_state=None):
 
                 # sparse matrices cannot multiply with 3d arrays so we use a loop over the modes                
                 for mi in xrange(self.model.number_of_modes):
-                    node_coupling[..., mi] = vertex_mapping * region_coupling[..., mi].T
+                    node_coupling[..., mi] = region_coupling[..., self.surface.region_mapping]
```
Comment (timpx) [inline tvb/simulator/simulator.py]: does not work with subcortical regions

File: tvb/simulator/simulator.py [hunk 2]
```diff
@@ -452,13 +452,10 @@ def __call__(self, simulation_length=None, random_state=None):
...
         if self.surface is not None:
-            # the vertex mapping array is huge but sparse.
-            # csr because I expect the row to have one value and I expect the dot to proceed row wise.
-            vertex_mapping = sparse.csr_matrix(self.surface.vertex_mapping)
-            # this is big a well. same shape as the vertex mapping.
+            # this is big a well. 
             region_average = sparse.csr_matrix(region_average)
             
-            node_coupling_shape = (vertex_mapping.shape[0], ncvar, self.model.number_of_modes)
+            node_coupling_shape = (self.surface.region_mapping.shape[0], ncvar, self.model.number_of_modes)
```
Comment (timpx) [inline tvb/simulator/simulator.py]: does not work with subcortical regions

**Comment by timpx** [comment]
some more work is needed for subcortical regions, because the shape of vertex mapping was number_of_cortical_regions + number_of_subcortical_regions; the vertex dealt with subcortical regions [coupling with](https://github.com/the-virtual-brain/tvb-library/blob/trunk/tvb/simulator/simulator.py#L477): 

```
for mi in xrange(self.model.number_of_modes):
    node_coupling[..., mi] = vertex_mapping * region_coupling[..., mi].T
```

 by giving the sum of all [subcortical regions activity](https://github.com/the-virtual-brain/tvb-library/blob/trunk/tvb/datatypes/cortex_scientific.py#L212):

```
vertex_mapping[self.number_of_vertices:, non_cortical_region_mapping] = 1
```

What does that mean, I don't know

**Comment by timpx** [comment]
So we have to decide how to deal with subcortical regions, self.region_sum should have the shape nb_of_cortical + nb_of_subcortical regions, so how do we inform cortex about the number of subcortical regions?

**Comment by maedoc** [comment]
@liadomide could we get a svn / git sync on tvb-library tomorrow so we can finish this up? I can handle it if you'd like.

**Comment by liadomide** [comment]
I've last synchronised tvb-library with svn yesterday. I see no change since then on svn or git. 

I am thinking that we could start fixing this from Cortex Datatype (not CorticalSurface) - meant to be a transient entity (not stored), and build when configuring a simulation (from GUI or from console).
The Cortex class should get a new name. @maedoc suggested Head.

Head will handle the geometry for the current stimulation (cortical surface, skin, skull surfaces, the region_mapping, the connectivity, sensors and projections). Some of the field enumerated above will be optional, but with a good flag in case they are missing (e.g. because the simulation is on regions level and not on surface). I can trace something in this direction tomorrow, but it would be great if one of you wants to take over for the rest of simulator required fixes.

**Comment by maedoc** [comment]
> last synchronised tvb-library with svn yesterday. I see no change since then on svn or git.

ok sounds good I just wanted to check.

> build when configuring a simulation

If I understand what you mean, the region mapping on disk would not contain subcortical info, but instead will be added on at runtime, so that we're sure the convention is always the same?

> cortical surface, skin, skull surfaces, the region_mapping, the connectivity, sensors and projections

this is a lot of stuff; I understand the (current) need for a transient God object, but ideally we could factor this a little more.

**Comment by maedoc** [comment]
@liadomide @timpx After re-reading I would prefer to work out the source spaces explicitly (i.e. make datatypes) and implement methods that handle their combination, rather than attempt to maintain these hacks. It can be implemented, tested, working this week, so I don't see the reason to not do it correctly now so we never come back to this issue again.

**Comment by liadomide** [comment]
ok

File: tvb/datatypes/cortex_scientific.py [hunk 3]
```diff
@@ -161,66 +161,16 @@ def _set_region_sum(self, spatial_mask):
...
-            cortical_region_mapping = [x for x in self.region_mapping if x in cortical_regions[0]]
-            non_cortical_region_mapping = [x for x in self.region_mapping if x in non_cortical_regions[0]]
-            vertex_mapping[numpy.arange(self.number_of_vertices), cortical_region_mapping] = 1
-            # the rest
-            vertex_mapping[self.number_of_vertices:, non_cortical_region_mapping] = 1
-
-        else:
-            vertex_mapping[numpy.arange(number_of_nodes), spatial_mask] = 1
-        self._vertex_mapping = vertex_mapping
+        self._region_sum = numpy.zeros((number_of_areas, number_of_nodes))
```
Comment (maedoc) [inline tvb/datatypes/cortex_scientific.py]: Region summation is better accomplished with `np.add.at(out, region_map, state)`: it's effectively sparse but generalizes to N-dimensions, also obliges use of pre-allocated output array.

File: tvb/simulator/simulator.py [hunk 4]
```diff
@@ -744,7 +739,7 @@ def _census_memory_requirement(self):
         try:
             memreq += self.surface.triangles.nbytes * 2
             memreq += self.surface.vertices.nbytes * 2
-            memreq += self.surface.vertex_mapping.nbytes * 4  # vertex_mapping, region_average, region_sum
+            memreq += self.surface.region_mapping.nbytes * self.number_of_nodes * 8. * 4  # region_average, region_sum
```
Comment (maedoc) [inline tvb/simulator/simulator.py:718]: ? Shouldn't be `region_mapping.nbytes + n * 8 * 4`?

In any case, memory estimation the simulator is fairly sloppy, I wouldn't spend time on this.

**Comment by maedoc** [comment]
@timpx can you

```
git remote add maedoc https://github.com/maedoc/tvb-library
git checkout trunk
git pull maedoc tim-trunk
git push
```

?

**Comment by timpx** [comment]
@maedoc done; is it working now?

**Comment by maedoc** [comment]
the basic surface simulation is running, but I haven't tested all the variations or loading the new lead fields. I'm cleaning up `Simulator.__call__` right now...

**Comment by maedoc** [comment]
We discussed a few conventions. One is that the region mapping's contents always refer to indices of the connectivity matrix. Some region mappings may contain only cortical regions even if the connectivity has subcortical, though it may contain those as well. A final convention is that the subcortical regions are at the end of the region mapping.

@timpx is this correct? Which of these assumptions did you want to relax? The third one, right?

**Comment by timpx** [comment]
Yes, convention 1 should be mandatory and be assumed everywhere in the code, while conventions 2 and 3 are optional, therefore should not be assumed in the code, but we could handle these cases for more flexibility and historical reasons.

**Comment by maedoc** [comment]
OK then we need to make a clever function to handle these. If the region mapping file contains only cortical indices, then 

``` python
setdiff1(r_[:number_of_regions], region_mapping)
```

is ok, but fails if the region mapping contains subcortical indices in which case we need

``` python
where(bincount(region_mapping) == 1)
```

**Comment by maedoc** [comment]
I think we can check which to use with `vertices.shape[0]==region_mapping.shape[0]`? So,

``` python
def clever_non_cortical_indices(surf, regmap, conn):
    if surf.vertices.shape[0] == regmap.shape: # doesn't have subcortical indices
        return setdiff1(r_[:conn.number_of_regions], regmap)
    else:
        return where(bincount(regmap)==1)[0]
```

Where to put such a function..?

**Comment by timpx** [comment]
+1 for bincount and +1 for setdiff1.
in cortex_scientific.py?

**Comment by maedoc** [comment]
will do. for now I've got the surf sim working with 192 (so subcortical) & new EEG sensors & projections. Seems to produce NaNs.. will have to see why.

Can you pull from `maedoc/tim-trunk`? Also, if you want to try, you need to update from svn and use svn's tvb_data.

**Comment by timpx** [comment]
done, i'll try this tomorrow... I mean today

**Comment by maedoc** [comment]
I put the other sensor files in the SVN as well, and I've pushed a [script](https://github.com/maedoc/tvb-library/blob/tim-trunk/tvb/simulator/demos/surface_stochastic_subcortical_seeg_eeg_meg.py) that runs with the nwe seeg meg and eeg sensors & projections. We'll still need to look at the signals to check they're good, but the code is in place.

@timpx can you pull these new changes, check it works for you, then we merge & sync with svn? @liadomide what do you think?

**Comment by maedoc** [comment]
I just pushed  a last commit on this branch, with an integration test for both cortical & subcortical cases, will all the new monitors. Can you merge that and then we merge this?

**Comment by maedoc** [comment]
[I know the EEG proj is producing NaNs](http://req.thevirtualbrain.org/browse/TVB-1800?focusedCommentId=14116&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14116) but it's not a code issue. We'll check the projections next week.

**Comment by maedoc** [comment]
I just merged this by hand

**Pull request closed without merging**